python/tvm/auto_scheduler/task_scheduler.py

```diff
@@ -0,0 +1,452 @@
+                    )
+                    assert grad <= 0
+                    gradients.append(grad)
+
+                if max(gradients) == min(gradients):
+                    task_idx = np.random.choice(len(gradients))
+                else:
+                    task_idx = np.argmin(gradients)
+            else:
+                raise ValueError("Invalid strategy: " + self.strategy)
+
+            self.tune_task(task_idx)
+            self.adjust_similarity_group(task_idx)
+
+    def tune_task(self, task_idx):
+        """Tune the select task for one round"""
+        if self.verbose >= 1:
+            logger.info("TaskScheduler: task id:\t%d", task_idx)
+        measure_inputs, measure_results = self.search_policies[task_idx].continue_search_one_round(
+            self.num_measures_per_round, self.measurer
+        )
+
+        for _, res in zip(measure_inputs, measure_results):
+            cost = array_mean(res.costs)
+            if cost < self.best_costs[task_idx]:
+                self.best_costs[task_idx] = cost
+
+        if len(measure_inputs) == 0:
+            self.dead_tasks.add(task_idx)
+
+        self.task_cts[task_idx] += 1
+        self.task_costs_history[task_idx].append(self.best_costs[task_idx])
+
+        self.ct += len(measure_inputs)
+        self.cur_score = self.compute_score(self.best_costs)
+
+        if self.verbose >= 1:
+            logger.info(
+                "TaskScheduler\tct: %d\testimated cost (ms): %.3f\ttime elapsed: %.2f\t"
+                "best_costs (ms): %s\ttask_ct: %s",
+                self.ct,
+                self.cur_score * 1e3,
+                time.time() - self.tic,
+                to_str_round(self.best_costs * 1e3, decimal=3),
+                self.task_cts,
+            )
+
+    def adjust_similarity_group(self, task_idx):
+        """adjust the similarity group for the selected task"""
+        group_id = self.tag_to_group_id.get(self.task_tags[task_idx], None)
+        if group_id is None or len(self.group_task_ids[group_id]) <= 1:
+            return
+
+        group_ids = self.group_task_ids[group_id]
+        best_group_flops = max([self.flop_cts[j] / self.best_costs[j] for j in group_ids])
+        cur_flops = self.flop_cts[task_idx] / self.best_costs[task_idx]
+
+        # if we tune a task for many times but it still cannot achieve
+        # a similar speed to the fastest one in its group, this means this task
+        # is actually not similar to other tasks in its group.
+        # So we will reomve it from its original group.
+        if cur_flops < best_group_flops / self.beta and self.task_cts[task_idx] > 5 + max(
+            self.task_cts[j] for j in group_ids if j != task_idx
+        ):
+            self.task_tags[task_idx] = None
+            group_ids.remove(task_idx)
+
+    def restore_status(self, log_file, num_measures_per_round):
+        """restore task_cts and best_costs from a log file"""
+        if log_file:
```

Comment:
Better to enforce `log_file` to be not None in this function. It seems meaningless to call this function with `log_file=None`.